### PR TITLE
fixes #21801; object field initialization with overloaded functions

### LINKED ISSUE #21801: Default object field initialization does not work for function types with overloaded functions
### Description

The following example does not compile with `nim -r main.nim`:
```Nim
func evaluate(i: int): float =
  0.0

func evaluate(): float =
  0.0

type SearchOptions = object
    evaluation: proc(): float = evaluate
```

However, this does:
```Nim
func evaluate(): float =
  0.0

type SearchOptions = object
    evaluation: proc(): float = evaluate
```

### Nim Version

Nim Compiler Version 1.9.3 [Linux: amd64]
Compiled at 2023-05-06
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: 85dbfc68b5c2020b32be9e3dc0d711b749fb5c6f
active boot switches: -d:release

### Current Output

```text
Hint: used config file '/home/tsoj/.choosenim/toolchains/nim-#devel/config/nim.cfg' [Conf]
Hint: used config file '/home/tsoj/.choosenim/toolchains/nim-#devel/config/config.nims' [Conf]
......................................................................
/tmp/testtest/main.nim(8, 33) Error: no tuple type for constructor
```

### Expected Output

```text
Should just compile and run, or if it correctly doesn't compile, then a better error message.
```

### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by ringabout**
fixes #21801

### CODE REVIEW

**Comment by ringabout** [comment]
CI is currently broken. It needs to wait for https://github.com/nim-lang/Nim/pull/21804

File: compiler/semtypes.nim [hunk 1]
```diff
@@ -227,6 +227,7 @@ proc isRecursiveType(t: PType, cycleDetector: var IntSet): bool =
 
 proc fitDefaultNode(c: PContext, n: PNode): PType =
   let expectedType = if n[^2].kind != nkEmpty: semTypeNode(c, n[^2], nil) else: nil
+  n[^1] = semExprWithType(c, n[^1], {efTypeAllowed}, expectedType)
```
Comment (Araq) [inline compiler/semtypes.nim]: Why the `efTypeAllowed` here? And isn't semchecking already done by `semConstExpr`?

Comment (ringabout) [reply compiler/semtypes.nim:230]: Seems that overloading resolving needs to be done by `semExpr`. `efTypeAllowed` is used to make the type mismatch error message better. I suppose I can remove it.

Comment (ringabout) [reply compiler/semtypes.nim]: You're right, I didn't notice that.

**Pull request merged**